aten/src/ATen/native/cuda/TensorTransformations.cu

```diff
@@ -0,0 +1,121 @@
+#include "ATen/native/TensorTransformations.h"
+
+#include "ATen/cuda/detail/IndexUtils.cuh"
+#include "ATen/NativeFunctions.h"
+#include "ATen/cuda/CUDATensorMethods.cuh"
+#include "ATen/cuda/CUDATypeConversion.cuh"
+
+namespace at {
+namespace native {
+
+#define AT_APPLY_THREADS_PER_BLOCK 32 * 16
+#define AT_APPLY_BLOCKS_PER_SM 4
+
+template <typename scalar_t, typename IndexType>
+#if __CUDA_ARCH__ >= 350
+__launch_bounds__(AT_APPLY_THREADS_PER_BLOCK, AT_APPLY_BLOCKS_PER_SM)
+#endif
+__global__ void
+kernel_pointwise_flip_apply2(const cuda::detail::TensorInfo<scalar_t, IndexType> in_tensor_info,
+                          cuda::detail::TensorInfo<scalar_t, IndexType> out_tensor_info,
+                          IndexType N,
+                          int flip_dim,
+                          int64_t total_dims) {
+  for (IndexType linear_index = blockIdx.x * blockDim.x + threadIdx.x; linear_index < N; linear_index += gridDim.x * blockDim.x) {
+    int64_t cur_indices = linear_index, rem = 0, dst_offset = 0;
+    for (int64_t i = 0; i < total_dims; i++) {
+      int64_t temp = cur_indices;
+      cur_indices = cur_indices / in_tensor_info.strides[i];
+      rem = temp - cur_indices * in_tensor_info.strides[i];
```

Comment:
You can restructure you code in such a way that there's one less divide - for the last iteration cur_indices is guaranteed to be less than the stride you are dividing it by (same applies to your second kernel).

Reply:
Great catch! I will change it accordingly.

Reply:
I am still a bit confused about the last iter, for example, if we have a tensor like this:
```
[
  [1, 2, 3],
  [4, 5, 6]
]
```
its sizes = (2, 3), strides = (3, 1). When linear_index = 2, at the last iter cur_indices = 2 > strides[-1] = 1